certdb/pg/pg.go

```diff
@@ -0,0 +1,209 @@
+	WHERE (serial = $2)`
+
+	insertOCSPSQL = `
+INSERT INTO ocsp_responses (serial, body, expiry)
+    VALUES ($1, $2, $3);`
+
+	updateOCSPSQL = `
+UPDATE ocsp_responses
+    SET expiry=$3, body=$2
+	WHERE (serial = $1);`
+
+	selectAllUnexpiredOCSPSQL = `
+SELECT %s FROM ocsp_responses
+WHERE current_timestamp < expiry;`
+
+	selectOCSPSQL = `
+SELECT %s FROM ocsp_responses
+    WHERE (serial = $1);`
+)
+
+func wrapCertStoreError(err error) error {
+	if err != nil {
+		return cferr.Wrap(cferr.CertStoreError, cferr.Unknown, err)
+	}
+	return nil
+}
+
+// InsertCertificate puts a certdb.CertificateRecord into db
+func InsertCertificate(db *sql.DB, cr *certdb.CertificateRecord) error {
+	res, err := db.Exec(
+		insertSQL,
+		cr.Serial,
+		cr.CALabel,
+		cr.Status,
+		cr.Reason,
+		cr.Expiry,
+		cr.RevokedAt,
+		cr.PEM,
+	)
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := res.RowsAffected()
+
+	if numRowsAffected != 1 {
+		return wrapCertStoreError(errors.New("Either zero or more than 1 rows are affected"))
+	}
+
+	return nil
+}
+
+// GetCertificate gets a certdb.CertificateRecord indexed by serial
+func GetCertificate(db *sql.DB, serial string) (*certdb.CertificateRecord, error) {
+	cr := new(certdb.CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectSQL, sqlstruct.Columns(*cr)), serial)
+	defer rows.Close()
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+
+	if rows.Next() {
+		return cr, wrapCertStoreError(sqlstruct.Scan(cr, rows))
+	}
+	return nil, nil
+}
+
+// GetUnexpiredCertificates gets all unexpired certificate from db
+func GetUnexpiredCertificates(db *sql.DB) (crs []*certdb.CertificateRecord, err error) {
+	cr := new(certdb.CertificateRecord)
+	rows, err := db.Query(fmt.Sprintf(selectAllUnexpiredSQL, sqlstruct.Columns(*cr)))
+	defer rows.Close()
+	if err != nil {
+		return nil, wrapCertStoreError(err)
+	}
+
+	for rows.Next() {
+		err = sqlstruct.Scan(cr, rows)
+		if err != nil {
+			return nil, wrapCertStoreError(err)
+		}
+		crs = append(crs, cr)
+	}
+
+	return crs, nil
+}
+
+// RevokeCertificate updates a certificate with a given serial number and marks it revoked
+func RevokeCertificate(db *sql.DB, serial string, reasonCode int) error {
+	result, err := db.Exec(updateRevokeSQL, reasonCode, serial)
+
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := result.RowsAffected()
+
+	if numRowsAffected != 1 {
+		return wrapCertStoreError(errors.New("Either zero or more than 1 rows are affected"))
+	}
+
+	return nil
+}
+
+// InsertOCSP puts a new OCSPRecord into the db
+func InsertOCSP(db *sql.DB, rr *certdb.OCSPRecord) error {
+	res, err := db.Exec(
+		insertOCSPSQL,
+		rr.Serial,
+		rr.Body,
+		rr.Expiry,
+	)
+	if err != nil {
+		return wrapCertStoreError(err)
+	}
+
+	numRowsAffected, _ := res.RowsAffected()
+
+	if numRowsAffected != 1 {
+		return wrapCertStoreError(errors.New("Either zero or more than 1 rows are affected"))
```

Comment:
Same comment as above:

I think this error should say how many rows were affected (or have a log.Error statement indicating how many rows were affected). golint normally complains that errors should begin with a lowercase letter, so that might be worth fixing here.